### ISSUE TITLE
Conflicting packages libapache2-mod-php8.1 php-fpm

### ISSUE THREAD

**Issue by joshcangit** [oerdnj/deb.sury.org#1710]
**Frequently asked questions**

- [x] I have read [Frequently Asked Questions](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions)

**Describe the bug**
Tried upgrading `php` to `php8.1` from `php8.0`.
Installing `php-fpm` and `php8.1-fpm` fails because `php` requires `libapache2-mod-php8.1` and `apache2`.
Both `php-fpm` and `libapache2-mod-php8.1` gets installed even though they conflict.
Was using `php-fpm` as connector to `nginx` so am I forced to purge and never install `nginx`?

**To Reproduce**
1. Add PPAs
```
sudo LC_ALL=en_US.UTF-8 add-apt-repository ppa:ondrej/php
sudo LC_ALL=en_US.UTF-8 add-apt-repository ppa:ondrej/php-qa
sudo add-apt-repository ppa:ondrej/apache2
```
2. Install `php` and `php-fpm`
```
sudo apt install php php-fpm
```
3. See error
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  apache2 apache2-bin apache2-data apache2-utils libapache2-mod-php8.1 libapr1 libaprutil1 libaprutil1-dbd-sqlite3 libaprutil1-ldap php-common php8.1 php8.1-cli php8.1-common php8.1-fpm php8.1-opcache
  php8.1-readline
Suggested packages:
  apache2-doc apache2-suexec-pristine | apache2-suexec-custom php-pear
The following NEW packages will be installed:
  apache2 apache2-bin apache2-data apache2-utils libapache2-mod-php8.1 libapr1 libaprutil1 libaprutil1-dbd-sqlite3 libaprutil1-ldap php php-common php-fpm php8.1 php8.1-cli php8.1-common php8.1-fpm
  php8.1-opcache php8.1-readline
0 upgraded, 18 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/8,413 kB of archives.
After this operation, 34.6 MB of additional disk space will be used.
Selecting previously unselected package libapr1:amd64.
(Reading database ... 264720 files and directories currently installed.)
Preparing to unpack .../00-libapr1_1.7.0-8+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking libapr1:amd64 (1.7.0-8+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package libaprutil1:amd64.
Preparing to unpack .../01-libaprutil1_1.6.1-5+ubuntu20.04.1+deb.sury.org+5_amd64.deb ...
Unpacking libaprutil1:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Selecting previously unselected package libaprutil1-dbd-sqlite3:amd64.
Preparing to unpack .../02-libaprutil1-dbd-sqlite3_1.6.1-5+ubuntu20.04.1+deb.sury.org+5_amd64.deb ...
Unpacking libaprutil1-dbd-sqlite3:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Selecting previously unselected package libaprutil1-ldap:amd64.
Preparing to unpack .../03-libaprutil1-ldap_1.6.1-5+ubuntu20.04.1+deb.sury.org+5_amd64.deb ...
Unpacking libaprutil1-ldap:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Selecting previously unselected package apache2-bin.
Preparing to unpack .../04-apache2-bin_2.4.52-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking apache2-bin (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package apache2-data.
Preparing to unpack .../05-apache2-data_2.4.52-2+ubuntu20.04.1+deb.sury.org+1_all.deb ...
Unpacking apache2-data (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package apache2-utils.
Preparing to unpack .../06-apache2-utils_2.4.52-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking apache2-utils (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package apache2.
Preparing to unpack .../07-apache2_2.4.52-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking apache2 (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php-common.
Preparing to unpack .../08-php-common_2%3a92+ubuntu20.04.1+deb.sury.org+2_all.deb ...
Unpacking php-common (2:92+ubuntu20.04.1+deb.sury.org+2) ...
Selecting previously unselected package php8.1-common.
Preparing to unpack .../09-php8.1-common_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking php8.1-common (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php8.1-opcache.
Preparing to unpack .../10-php8.1-opcache_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking php8.1-opcache (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php8.1-readline.
Preparing to unpack .../11-php8.1-readline_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking php8.1-readline (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php8.1-cli.
Preparing to unpack .../12-php8.1-cli_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking php8.1-cli (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package libapache2-mod-php8.1.
Preparing to unpack .../13-libapache2-mod-php8.1_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking libapache2-mod-php8.1 (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php8.1-fpm.
Preparing to unpack .../14-php8.1-fpm_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_amd64.deb ...
Unpacking php8.1-fpm (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php8.1.
Preparing to unpack .../15-php8.1_8.1.1-2+ubuntu20.04.1+deb.sury.org+1_all.deb ...
Unpacking php8.1 (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Selecting previously unselected package php.
Preparing to unpack .../16-php_2%3a8.1+92+ubuntu20.04.1+deb.sury.org+2_all.deb ...
Unpacking php (2:8.1+92+ubuntu20.04.1+deb.sury.org+2) ...
Selecting previously unselected package php-fpm.
Preparing to unpack .../17-php-fpm_2%3a8.1+92+ubuntu20.04.1+deb.sury.org+2_all.deb ...
Unpacking php-fpm (2:8.1+92+ubuntu20.04.1+deb.sury.org+2) ...
Setting up php-common (2:92+ubuntu20.04.1+deb.sury.org+2) ...
Created symlink /etc/systemd/system/timers.target.wants/phpsessionclean.timer → /lib/systemd/system/phpsessionclean.timer.
Setting up php8.1-common (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...

Creating config file /etc/php/8.1/mods-available/calendar.ini with new version

Creating config file /etc/php/8.1/mods-available/ctype.ini with new version

Creating config file /etc/php/8.1/mods-available/exif.ini with new version

Creating config file /etc/php/8.1/mods-available/fileinfo.ini with new version

Creating config file /etc/php/8.1/mods-available/ffi.ini with new version

Creating config file /etc/php/8.1/mods-available/ftp.ini with new version

Creating config file /etc/php/8.1/mods-available/gettext.ini with new version

Creating config file /etc/php/8.1/mods-available/iconv.ini with new version

Creating config file /etc/php/8.1/mods-available/pdo.ini with new version

Creating config file /etc/php/8.1/mods-available/phar.ini with new version

Creating config file /etc/php/8.1/mods-available/posix.ini with new version

Creating config file /etc/php/8.1/mods-available/shmop.ini with new version

Creating config file /etc/php/8.1/mods-available/sockets.ini with new version

Creating config file /etc/php/8.1/mods-available/sysvmsg.ini with new version

Creating config file /etc/php/8.1/mods-available/sysvsem.ini with new version

Creating config file /etc/php/8.1/mods-available/sysvshm.ini with new version

Creating config file /etc/php/8.1/mods-available/tokenizer.ini with new version
Setting up libapr1:amd64 (1.7.0-8+ubuntu20.04.1+deb.sury.org+1) ...
Setting up php8.1-readline (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...

Creating config file /etc/php/8.1/mods-available/readline.ini with new version
Setting up apache2-data (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Setting up php8.1-opcache (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...

Creating config file /etc/php/8.1/mods-available/opcache.ini with new version
Setting up libaprutil1:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Setting up libaprutil1-ldap:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Setting up libaprutil1-dbd-sqlite3:amd64 (1.6.1-5+ubuntu20.04.1+deb.sury.org+5) ...
Setting up php8.1-cli (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
update-alternatives: using /usr/bin/php8.1 to provide /usr/bin/php (php) in auto mode
update-alternatives: using /usr/bin/phar8.1 to provide /usr/bin/phar (phar) in auto mode
update-alternatives: using /usr/bin/phar.phar8.1 to provide /usr/bin/phar.phar (phar.phar) in auto mode

Creating config file /etc/php/8.1/cli/php.ini with new version
Setting up apache2-utils (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Setting up apache2-bin (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Setting up php8.1-fpm (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Package apache2 is not configured yet. Will defer actions by package php8.1-fpm.

Creating config file /etc/php/8.1/fpm/php.ini with new version
NOTICE: Not enabling PHP 8.1 FPM by default.
NOTICE: To enable PHP 8.1 FPM in Apache2 do:
NOTICE: a2enmod proxy_fcgi setenvif
NOTICE: a2enconf php8.1-fpm
NOTICE: You are seeing this message because you have apache2 package installed.
Created symlink /etc/systemd/system/multi-user.target.wants/php8.1-fpm.service → /lib/systemd/system/php8.1-fpm.service.
Job for php8.1-fpm.service failed because the control process exited with error code.
See "systemctl status php8.1-fpm.service" and "journalctl -xe" for details.
invoke-rc.d: initscript php8.1-fpm, action "restart" failed.
● php8.1-fpm.service - The PHP 8.1 FastCGI Process Manager
     Loaded: loaded (/lib/systemd/system/php8.1-fpm.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2022-01-18 10:58:42 +08; 8ms ago
       Docs: man:php-fpm8.1(8)
    Process: 215945 ExecStart=/usr/sbin/php-fpm8.1 --nodaemonize --fpm-config /etc/php/8.1/fpm/php-fpm.conf (code=exited, status=127)
    Process: 215946 ExecStopPost=/usr/lib/php/php-fpm-socket-helper remove /run/php/php-fpm.sock /etc/php/8.1/fpm/pool.d/www.conf 81 (code=exited, status=0/SUCCESS)
   Main PID: 215945 (code=exited, status=127)

Jan 18 10:58:42 joshua-ThinkPad systemd[1]: Starting The PHP 8.1 FastCGI Process Manager...
Jan 18 10:58:42 joshua-ThinkPad php-fpm8.1[215945]: /usr/sbin/php-fpm8.1: symbol lookup error: /usr/sbin/php-fpm8.1: undefined symbol: crypto_core_ristretto255_scalar_invert
Jan 18 10:58:42 joshua-ThinkPad systemd[1]: php8.1-fpm.service: Main process exited, code=exited, status=127/n/a
Jan 18 10:58:42 joshua-ThinkPad systemd[1]: php8.1-fpm.service: Failed with result 'exit-code'.
Jan 18 10:58:42 joshua-ThinkPad systemd[1]: Failed to start The PHP 8.1 FastCGI Process Manager.
dpkg: error processing package php8.1-fpm (--configure):
 installed php8.1-fpm package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of php-fpm:
 php-fpm depends on php8.1-fpm; however:
  Package php8.1-fpm is not configured yet.

dpkg: error processing package php-fpm (--configure):
 dependency problems - leaving unconfigured
Setting up libapache2-mod-php8.1 (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
No apport report written because the error message indicates its a followup error from a previous failure.
                                                                                                          Package apache2 is not configured yet. Will defer actions by package libapache2-mod-php8.1.

Creating config file /etc/php/8.1/apache2/php.ini with new version
No module matches 
Setting up apache2 (2.4.52-2+ubuntu20.04.1+deb.sury.org+1) ...
Enabling module mpm_event.
Enabling module authz_core.
Enabling module authz_host.
Enabling module authn_core.
Enabling module auth_basic.
Enabling module access_compat.
Enabling module authn_file.
Enabling module authz_user.
Enabling module alias.
Enabling module dir.
Enabling module autoindex.
Enabling module env.
Enabling module mime.
Enabling module negotiation.
Enabling module setenvif.
Enabling module filter.
Enabling module deflate.
Enabling module status.
Enabling module reqtimeout.
Enabling conf charset.
Enabling conf localized-error-pages.
Enabling conf other-vhosts-access-log.
Enabling conf security.
Enabling conf serve-cgi-bin.
Enabling site 000-default.
info: Switch to mpm prefork for package libapache2-mod-php8.1
Module mpm_event disabled.
Enabling module mpm_prefork.
info: Executing deferred 'a2enmod php8.1' for package libapache2-mod-php8.1
Enabling module php8.1.
Created symlink /etc/systemd/system/multi-user.target.wants/apache2.service → /lib/systemd/system/apache2.service.
Job for apache2.service failed because the control process exited with error code.
See "systemctl status apache2.service" and "journalctl -xe" for details.
invoke-rc.d: initscript apache2, action "restart" failed.
● apache2.service - The Apache HTTP Server
     Loaded: loaded (/lib/systemd/system/apache2.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2022-01-18 10:58:50 +08; 9ms ago
       Docs: https://httpd.apache.org/docs/2.4/
    Process: 217073 ExecStart=/usr/sbin/apachectl start (code=exited, status=1/FAILURE)

Jan 18 10:58:50 joshua-ThinkPad systemd[1]: Starting The Apache HTTP Server...
Jan 18 10:58:50 joshua-ThinkPad apachectl[217076]: apache2: Syntax error on line 146 of /etc/apache2/apache2.conf: Syntax error on line 3 of /etc/apache2/mods-enabled/php8.1.load: Cannot load /usr/lib/ap
ache2/modules/libphp8.1.so into server: /usr/lib/apache2/modules/libphp8.1.so: undefined symbol: crypto_core_ristretto255_scalar_complement
Jan 18 10:58:50 joshua-ThinkPad apachectl[217073]: Action 'start' failed.
Jan 18 10:58:50 joshua-ThinkPad apachectl[217073]: The Apache error log may have more information.
Jan 18 10:58:50 joshua-ThinkPad systemd[1]: apache2.service: Control process exited, code=exited, status=1/FAILURE
Jan 18 10:58:50 joshua-ThinkPad systemd[1]: apache2.service: Failed with result 'exit-code'.
Jan 18 10:58:50 joshua-ThinkPad systemd[1]: Failed to start The Apache HTTP Server.
Setting up php8.1 (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Setting up php (2:8.1+92+ubuntu20.04.1+deb.sury.org+2) ...
Processing triggers for systemd (245.4-4ubuntu3.15) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for ureadahead (0.100.0-21) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Processing triggers for php8.1-cli (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Processing triggers for libapache2-mod-php8.1 (8.1.1-2+ubuntu20.04.1+deb.sury.org+1) ...
Errors were encountered while processing:
 php8.1-fpm
 php-fpm
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Expected behavior**
I expected installing `php` and `php-fpm` that `libapache2-mod-php8.1` does not also get installed.

**Distribution (please complete the following information):**
 - OS: Ubuntu 20.04.3 LTS
 - Architecture: amd64
 - Repository: Ubuntu PPA

**Package(s) (please complete the following information):**
```
php:
  Installed: 2:8.1+92+ubuntu20.04.1+deb.sury.org+2
  Candidate: 2:8.1+92+ubuntu20.04.1+deb.sury.org+2
  Version table:
 *** 2:8.1+92+ubuntu20.04.1+deb.sury.org+2 500
        500 http://ppa.launchpad.net/ondrej/php/ubuntu focal/main amd64 Packages
        500 http://ppa.launchpad.net/ondrej/php/ubuntu focal/main i386 Packages
        100 /var/lib/dpkg/status
     2:8.0+85+ubuntu20.04.1+deb.sury.org+1 500
        500 http://ppa.launchpad.net/ondrej/php-qa/ubuntu focal/main amd64 Packages
        500 http://ppa.launchpad.net/ondrej/php-qa/ubuntu focal/main i386 Packages
     2:7.4+75 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal/main i386 Packages
```
```
php-fpm:
  Installed: 2:8.1+92+ubuntu20.04.1+deb.sury.org+2
  Candidate: 2:8.1+92+ubuntu20.04.1+deb.sury.org+2
  Version table:
 *** 2:8.1+92+ubuntu20.04.1+deb.sury.org+2 500
        500 http://ppa.launchpad.net/ondrej/php/ubuntu focal/main amd64 Packages
        500 http://ppa.launchpad.net/ondrej/php/ubuntu focal/main i386 Packages
        100 /var/lib/dpkg/status
     2:8.0+85+ubuntu20.04.1+deb.sury.org+1 500
        500 http://ppa.launchpad.net/ondrej/php-qa/ubuntu focal/main amd64 Packages
        500 http://ppa.launchpad.net/ondrej/php-qa/ubuntu focal/main i386 Packages
     2:7.4+75 500
        500 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal/universe i386 Packages
```
```
libapache2-mod-php8.1:
  Installed: 8.1.1-2+ubuntu20.04.1+deb.sury.org+1
  Candidate: 8.1.1-2+ubuntu20.04.1+deb.sury.org+1
  Version table:
 *** 8.1.1-2+ubuntu20.04.1+deb.sury.org+1 500
        500 http://ppa.launchpad.net/ondrej/php/ubuntu focal/main amd64 Packages
        100 /var/lib/dpkg/status
     8.1.0~beta3-1+ubuntu20.04.1+deb.sury.org+1 500
        500 http://ppa.launchpad.net/ondrej/php-qa/ubuntu focal/main amd64 Packages
```

**Additional context**
Using nginx.org mainline repository not PPA.
```
nginx:
  Installed: 1.21.5-1~focal
  Candidate: 1.21.5-1~focal
  Version table:
 *** 1.21.5-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
        100 /var/lib/dpkg/status
     1.21.4-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.21.3-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.21.2-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.21.1-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.21.0-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.10-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.9-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.8-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.7-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.6-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.5-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.4-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.3-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.2-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.1-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.19.0-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.18.0-0ubuntu1.2 500
        500 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu focal-security/main i386 Packages
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal-updates/main i386 Packages
     1.17.10-1~focal 500
        500 http://nginx.org/packages/mainline/ubuntu focal/nginx amd64 Packages
     1.17.10-0ubuntu1 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal/main i386 Packages
```

**Comment by oerdnj** [comment]
> Installing php-fpm and php8.1-fpm fails because php requires libapache2-mod-php8.1 and apache2.

No, it does not, and you haven't read FAQ.

> Both php-fpm and libapache2-mod-php8.1 gets installed even though they conflict.

No, they don't conflict. They can't be enabled at the same time in the apache2.

> Was using php-fpm as connector to nginx so am I forced to purge and never install nginx?

No, the problem is different. You've just jumped to conclusions without reading the logs. It's all in there.

**Comment by joshcangit** [comment]
Sorry, didn't fully read the FAQ.
Turns out I was wrong, installing the `php` package only made the problem worse.
You're right. It is a different problem.

**Comment by joshcangit** [comment]
I solved it.
Turns out there was the file `libsodium.so.23.1.0` and its symlink `libsodium.so.23` inside `/usr/local/lib`.
Those were probably from a previous package installation.
By deleting those 2 files, php stop throwing that error.
